client/api/src/notifications.rs

```diff
@@ -114,10 +156,101 @@ impl<Block: BlockT> Default for StorageNotifications<Block> {
 	}
 }
 
+struct PrintKeys<'a>(&'a Option<HashSet<StorageKey>>);
+impl<'a> std::fmt::Display for PrintKeys<'a> {
+	fn fmt(&self, fmt: &mut std::fmt::Formatter) -> std::fmt::Result {
+		if let Some(keys) = self.0.as_ref() {
+			write!(fmt, "[")?;
+			let mut is_first = true;
+			for key in keys {
+				if is_first {
+					is_first = false;
+				} else {
+					write!(fmt, ", ")?;
+				}
+				write!(fmt, "{}", HexDisplay::from(key))?;
+			}
+			write!(fmt, "]")
+		} else {
+			write!(fmt, "None")
+		}
+	}
+}
+
+struct PrintChildKeys<'a>(&'a Option<HashMap<StorageKey, Option<HashSet<StorageKey>>>>);
+impl<'a> std::fmt::Display for PrintChildKeys<'a> {
+	fn fmt(&self, fmt: &mut std::fmt::Formatter) -> std::fmt::Result {
+		if let Some(child_keys) = self.0.as_ref() {
+			write!(fmt, "{{")?;
+			let mut is_first = true;
+			for (key, values) in child_keys {
+				if is_first {
+					is_first = false;
+				} else {
+					write!(fmt, ", ")?;
+				}
+				write!(fmt, "{}={}", HexDisplay::from(key), PrintKeys(values))?;
+			}
+			write!(fmt, "}}")
+		} else {
+			write!(fmt, "None")
+		}
+	}
+}
+
 impl<Block: BlockT> StorageNotifications<Block> {
 	/// Initialize a new StorageNotifications
 	/// optionally pass a prometheus registry to send subscriber metrics to
 	pub fn new(prometheus_registry: Option<Registry>) -> Self {
+		StorageNotifications(Arc::new(Mutex::new(StorageNotificationsImpl::new(
+			prometheus_registry,
+		))))
+	}
+
+	/// Trigger notification to all listeners.
+	///
+	/// Note the changes are going to be filtered by listener's filter key.
+	/// In fact no event might be sent if clients are not interested in the changes.
+	pub fn trigger(
+		&mut self,
+		hash: &Block::Hash,
+		changeset: impl Iterator<Item = (Vec<u8>, Option<Vec<u8>>)>,
+		child_changeset: impl Iterator<
+			Item = (Vec<u8>, impl Iterator<Item = (Vec<u8>, Option<Vec<u8>>)>),
+		>,
+	) {
+		self.0.lock().trigger(hash, changeset, child_changeset);
+	}
+
+	/// Start listening for particular storage keys.
+	pub fn listen(
+		&mut self,
+		filter_keys: Option<&[StorageKey]>,
+		filter_child_keys: Option<&[(StorageKey, Option<Vec<StorageKey>>)]>,
+	) -> StorageEventStream<Block::Hash> {
+		let (id, rx) = self.0.lock().listen(filter_keys, filter_child_keys);
+		let storage_notifications = Arc::downgrade(&self.0);
+		StorageEventStream {
+			rx,
+			unsubscribe: Some(Box::new(move |was_triggered| {
+				if let Some(storage_notifications) = storage_notifications.upgrade() {
+					if !was_triggered {
```

Comment:
I would handle this directly in the `StorageEventStream`

Reply:
I did it this way to not have to change the signature of `StorageEventStream`. To call `remove_subscriber` we need a reference to `StorageNotificationsImpl<Block>`, while the `StorageEventStream` takes only `Block::Hash` as its generic parameter instead of the whole `Block`, so either the signature needs to change, or it needs to be done indirectly like I did, and I didn't want to introduce more churn than it's necessary so I picked the second option.

But now that I think about it I guess I can change `StorageNotificationsImpl` to take `Hash` since it's a private type anyway, and then I can indeed move this directly to `StorageEventStream`.